ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlGenericRowAvroDeserializerTest.java

```diff
@@ -201,4 +213,342 @@ public void shouldDeserializeWithMissingFields() {
+        .intType().and()
+        .stringType()
+        .endUnion();
+    final Schema ksqlSchema = SchemaBuilder.struct()
+        .field("int", Schema.OPTIONAL_INT32_SCHEMA)
+        .field("string", Schema.OPTIONAL_STRING_SCHEMA)
+        .optional()
+        .build();
+    final Struct ksqlValue = new Struct(ksqlSchema).put("string", "foobar");
+    shouldDeserializeTypeCorrectly(avroSchema, "foobar", ksqlSchema, ksqlValue);
+  }
+
+  private void shouldDeserializeConnectTypeCorrectly(final Schema connectSchema,
+                                                     final Object connectValue,
+                                                     final Schema ksqlSchema,
+                                                     final Object ksqlValue) {
+    final SchemaRegistryClient schemaRegistryClient = new MockSchemaRegistryClient();
+
+    final Schema connectRecordSchema = SchemaBuilder.struct()
+        .field("field0", connectSchema)
+        .build();
+    final Struct connectRecord = new Struct(connectRecordSchema);
+    connectRecord.put("field0", connectValue);
+
+    final AvroConverter converter = new AvroConverter(schemaRegistryClient);
+    converter.configure(
+        ImmutableMap.of(
+            AbstractKafkaAvroSerDeConfig.AUTO_REGISTER_SCHEMAS, true,
+            AbstractKafkaAvroSerDeConfig.SCHEMA_REGISTRY_URL_CONFIG, ""
+        ),
+        false
+    );
+
+    final byte[] bytes = converter.fromConnectData("topic", connectRecordSchema, connectRecord);
+
+    final Schema ksqlRecordSchema = SchemaBuilder.struct()
+        .field("field0", ksqlSchema)
+        .optional()
+        .build();
+
+    final Deserializer<GenericRow> deserializer =
+        new KsqlAvroTopicSerDe().getGenericRowSerde(
+            ksqlRecordSchema, ksqlConfig, false, schemaRegistryClient).deserializer();
+
+    final GenericRow row = deserializer.deserialize("topic", bytes);
+
+    assertThat(row.getColumns().size(), equalTo(1));
+    assertThat(row.getColumns().get(0), equalTo(ksqlValue));
+  }
+
+  @Test
+  public void shouldDeserializeConnectInt8ToInteger() {
+    shouldDeserializeConnectTypeCorrectly(
+        Schema.INT8_SCHEMA,
+        (byte) 32,
+        Schema.OPTIONAL_INT32_SCHEMA,
+        32
+    );
+  }
+
+  @Test
+  public void shouldDeserializeConnectInt16ToInteger() {
+    shouldDeserializeConnectTypeCorrectly(
+        Schema.INT16_SCHEMA,
+        (short) 16384,
+        Schema.OPTIONAL_INT32_SCHEMA,
+        16384
+    );
+  }
+
+  @Test
+  public void shouldDeserializeConnectInt8ToBigint() {
+    shouldDeserializeConnectTypeCorrectly(
+        Schema.INT8_SCHEMA,
+        (byte) 32,
+        Schema.OPTIONAL_INT64_SCHEMA,
+        32L
+    );
+  }
+
+  @Test
+  public void shouldDeserializeConnectInt16ToBigint() {
+    shouldDeserializeConnectTypeCorrectly(
+        Schema.INT16_SCHEMA,
+        (short) 16384,
+        Schema.OPTIONAL_INT64_SCHEMA,
+        16384L
+    );
+  }
```

Comment:
A test to validate the correct column name in error message would be nice to have.

Reply:
This was added to ConnectDataTranslatortest